teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+  recommends we prefer that when we need to pull in some other language, we use existing languages which are widely
+  used and supported. Instead of being an early adopter of CUE we could wait until it is more popular and then consider
+  using it (we can also delay this decision until we want to add more schema support).
+
+### Matching interfaces and extra keys
+
+An object provided by one Task (as a result) will be considered to match the object expected by another Task (as a
+params) if the object in the result contains _at least_ the [required](#required-vs-optional-keys) keys declared by the
+param in the consuming Task. The extra keys will be ignored, allowing us to take a
+[duck typing](https://en.wikipedia.org/wiki/Duck_typing) approach and maximize interoperabiltiy between tasks.
+
+For example imagine a Task which requires a param object that contains `url` and `httpsProxy`. Another task produces an
+object with those two keys in addition to other keys. It should be possible to pass the object with the additional keys
+directly to the Task which requires only `url` and `httpsProxy` without needing to modify the object in between.
+
+If we expand our JSON Schema support in the future (and we have use cases to support it) we could allow Tasks to express
+if they would like to override this and not allow additional keys to be provided via
+[JSON Schema's `additionalProperties` field](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties)
+(which defaults to allowing additional properties).
+
+**When the Task writes more keys than it declares in its results:**
+
+- The TaskRun will succeed
+- The additional keys will be included in the TaskRun status
+- The additional keys will not be available for variable replacement in a Pipeline; only the declared keys will be
+  available
+
+**When a Task uses the result of a previous Task which declares more keys in addition to what the Task needs:**
+
+- This will be allowed
+- The TaskRun will be created with only the keys that the Task needs; the rest will be ignored
+
+### Required vs optional keys
+
+We have several options for how we define and handle required and optional keys:
+
+1. Embrace the optional by default behavior of JSON Schema; instead of validating the presences of "required" keys via
+   jsonschema, we could do the validation based on the declared variable replacement, e.g. say a Task declares it needs
+   an object called `foo` with an attribute called `bar`. At runtime if the Task is provided with an object
+   `foo` that doesn't have an attribute called `bar` this will be okay unless the Task contains variable replacement
+   that uses it (`$(params.foo.bar)`). Con: weird because effectively we're implying `required` (assuming the fields are
+   used, and if they aren't used, why are they there) even though the JSON Schema declaration says the fields are
+   optional.
```

Comment:
i would really like this to be the default, my main hesitations are:

* this means having a different default from jsonschema (https://json-schema.org/understanding-json-schema/reference/object.html#required-properties)
* i think @wlynch made a good point about how optional by default would interact well with shared schemas: https://github.com/tektoncd/community/pull/479#discussion_r692373651

Reply:
sounds good, now I'm wondering if we should close issue https://github.com/tektoncd/pipeline/issues/3497 and document the existing behavior because it's consistent with what we'll be doing here?

Reply:
left a comment on the issue: https://github.com/tektoncd/pipeline/issues/3497#issuecomment-1029570957

Reply:
kk, re. tektoncd/pipeline#3497 it makes sense but i feel a bit sad about it XD - im happy with whichever we decide, and at least once we decide we can be clear and consistent about it

Reply:
Thinking about this more, I'm back to thinking that it's important Tasks declare an interface that consumers rely on, and so I would rather fail when expected results are not present - and maybe that should apply be default to keys as well (more about this at https://github.com/tektoncd/pipeline/issues/3497#issuecomment-1031998997)

I've got an additional option to add to our list:

6. Add a field to each result declaration that is defined entirely by us, e.g. `keysRequiredByDefault` which defaults to true and causes us to treat all keys in an object schema declaration as required by default (i.e. literally treat each as required when evaluating json schema) unless the value is explicitly set to false @_@

I'm not sure I like it very much but maybe something like that is the only way to get the guarantees we need to Task consumers to have, balanced with using JSON schema as it is specified @_@